lib/internal/modules/esm/module_job.js

```diff
@@ -151,7 +167,30 @@ class ModuleJob {
     await this.instantiate();
     const timeout = -1;
     const breakOnSigint = false;
-    await this.module.evaluate(timeout, breakOnSigint);
+    try {
+      await this.module.evaluate(timeout, breakOnSigint);
+    } catch (e) {
+      if (e.name === 'ReferenceError' &&
+          StringPrototypeEndsWith(e.message, ' is not defined') &&
+          isCommonJSGlobalLikeNotDefinedError(e.message)) {
+        e.message += ' in ES module scope';
+
+        if (StringPrototypeStartsWith(e.message, 'require')) {
+          e.message += '. You can use `await import` instead';
+        }
+
+        if (StringPrototypeEndsWith(this.module.url, '.js') &&
+            StringPrototypeStartsWith(this.module.url, 'file://')) {
+          e.message +=
+            '. It seems you are trying to load a file using `.js` extension ' +
+            'inside a folder containing a `package.json` that includes ' +
+            '`"type": "module"`; you need to use the `.cjs` extension to ' +
+            'load the file as a CommonJS module, or add a `package.json` ' +
+            'file with ``"type": "commonjs"`.';
```

Comment:
This is way too long and reminds me of clippy. Something more short and to the point would be better:

```suggestion
			'CommonJS files can be loaded using the package.json type option or .cjs'
```

![](https://upload.wikimedia.org/wikipedia/en/d/db/Clippy-letter.PNG)

Reply:
Note that it's not much longer than the error message when using named import with CJS: https://github.com/nodejs/node/blob/5318e53fd89db5ba90036d040cb0558060813f18/lib/internal/modules/esm/module_job.js#L128-L134

In the linked issue, we have a user reporting they got such error without realising they were under a `"module"` package scope, having a very explicit error message would help I think. (and also everyone loved Clippy anyway, right? right? 🙃). Anyway, if the consensus is to have a shorter message, I'll take your suggestion.

Reply:
I was just about to say, we should make this similar to the “named export” error message, which includes the path to the controlling `package.json` file and things like that. An error message that includes the context to explain what’s going on seems worth the added length. Something like:

```
your-file-here.js cannot be parsed as an ES module, but it is being treated as an ES module
because it has a .js extension and /Users/devsnek/projects/foo/package.json contains
"type": "module". To treat it as a CommonJS file, rename it to your-file-here.cjs.
```

I’m not sure we want to recommend changing the relevant `"type"` value, since that would likely affect other files that might not be CommonJS. We could also do without the second sentence in my example above, though then we’re leaving the user to Google the error message to figure out what to do to fix this. (The other potential solution that comes to mind is to refactor the offending file into ESM, but that seems a bit much for an error message to suggest.)

Reply:
Alternatively, we could link to the docs rather than including the second sentence, and the docs are able to be much more precise without having to be quite so brief.

Reply:
I don't think we do link to our docs in any error message so far, maybe there's a good reason for not doing it? If there's none, that seems ideal indeed.